lib/core/utils/get-selector.js

```diff
@@ -1,38 +1,176 @@
 const escapeSelector = axe.utils.escapeSelector;
 let isXHTML;
+const ignoredAttributes = [
+  'class', 'style', 'id',
+  'selected', 'checked',
+  'disabled', 'tabindex',
+  'aria-checked', 'aria-selected',
+  'aria-invalid', 'aria-activedescendant',
+  'aria-busy', 'aria-disabled', 'aria-expanded',
+  'aria-grabbed', 'aria-pressed', 'aria-valuenow'
+  ];
+const MAXATTRIBUTELENGTH = 31;
 
-function isUncommonClassName (className) {
-  return ![
-    'focus', 'hover',
-    'hidden', 'visible',
-    'dirty', 'touched', 'valid', 'disable',
-    'enable', 'active', 'col-'
-  ].find(str => className.includes(str));
+/**
+ * get the attribute name and value as a string
+ * @param {Attribute} at    The attribute
+ * @return {String}
+ */
+function getAttributeNameValue(node, at) {
+  const name = at.name;
+  let atnv;
+  
+  if (name.indexOf('href') !== -1 || name.indexOf('src') !== -1) {
+      let value = encodeURI(axe.utils.getFriendlyUriEnd(node.getAttribute(name)));
+      if (value) {
+        atnv = escapeSelector(at.name) + '$="' + value + '"';  
+      } else {
+        return;
+      }
+  } else {
+    atnv = escapeSelector(name) + '="' + escapeSelector(at.value) + '"';
+  }
+  return atnv;
 }
 
-function getDistinctClassList (elm) {
-  if (!elm.classList || elm.classList.length === 0) {
-    return [];
-  }
+/**
+ * Filter the attributes
+ * @param {Attribute}    The potential attribute
+ * @return {Boolean}     Whether to include or exclude
+ */
+function filterAttributes(at) {
+  return !ignoredAttributes.includes(at.name) && at.name.indexOf(':') === -1 &&
+    (!at.value || at.value.length < MAXATTRIBUTELENGTH);
+}
 
-  const siblings = elm.parentNode && Array.from(elm.parentNode.children || '') || [];
-  return siblings.reduce((classList, childElm) => {
-    if (elm === childElm) {
-      return classList;
-    } else {
-      return classList.filter(classItem => {
-        return !childElm.classList.contains(classItem);
-      });
+/**
+ * Calculate the statistics for the classes, attributes and tags on the page, using
+ * the virtual DOM tree
+ * @param {Object} domTree    The root node of the virtual DOM tree
+ * @returns {Object}          The statistics consisting of three maps, one for classes,
+ *                            one for tags and one for attributes. The map values are
+ *                            the counts for how many elements with that feature exist
+ */
+axe.utils.getSelectorData = function (domTree) {
```

Comment:
I found a flaw in this method while thinking about checking if IDs are unique. We're not taking into account that our selectors are running in different trees. The count we're building up here is across the flattened tree. A class may be unique in its own shadow tree, while being fairly common across the flattened tree. We actually don't need to know how common something is across the flattened tree. The thing that matters is how common it is in its own light/shadow DOM tree.

I don't think it's difficult to fix. We can make a different selectorData object for each tree. We already track what tree we're looking at with the shadowId property. If we pass the selectorData object around instead of using a single global, it should be pretty easy to change this around.

Reply:
Yes but the idea is not that these stats are perfect but rather that they are better than hard coded ones.

What is the likelihood of clashes? Pretty small.

So the most likely downside of the current implementation is that the selector may appear to be more unique than it actually is, which may mean that the tag gets used a bit less frequently than it should. The performance impact of that is minimal. I think the simplicity of the code is worth the potential downside.